cub/warp/warp_exchange.cuh

```diff
@@ -468,5 +468,89 @@ public:
   //@}  end member group
 };
 
+template <typename InputT, int ITEMS_PER_THREAD>
+class WarpExchangeRegister
+{
+  static_assert(PowerOfTwo<ITEMS_PER_THREAD>::VALUE, "ITEMS_PER_THREAD must be a power of two");
+
+  template <int SIZE, int idx>
+  class CompileTimeArrayImpl; // fwd decl for partial specializations
+
+  template <int SIZE> // terminating partial specialization
+  class CompileTimeArrayImpl<SIZE, SIZE>
+  {
+  protected:
+    static_assert(PowerOfTwo<SIZE>::VALUE, "SIZE must be a power of two");
+
+    // used for dumping back the individual values after transposing
+    InputT (&output_items)[SIZE];
+
+    template <int>
+    __device__ void Loop(bool, unsigned)
+    {}
+
+  public:
+    __device__ CompileTimeArrayImpl(InputT (&input_items)[SIZE])
+        : output_items{input_items}
+    {}
+  };
+
+  template <int SIZE, int idx> // concrete recursion class
+  class CompileTimeArrayImpl : protected CompileTimeArrayImpl<SIZE, idx + 1>
+  {
+  protected:
+    static_assert(PowerOfTwo<SIZE>::VALUE, "SIZE must be a power of two");
+
+    InputT val;
+
+    template <int NUM_ENTRIES>
+    __device__ void Loop(const bool xor_bit_set, const unsigned mask)
+    {
+      // NOTE: Do *NOT* try to refactor this code to use a reference, since nvcc
+      //       tends to choke on it and then drop everything into local memory.
+      const InputT send_val = (xor_bit_set ? CompileTimeArrayImpl<SIZE, idx>::val
+                                           : CompileTimeArrayImpl<SIZE, idx + NUM_ENTRIES>::val);
+      const InputT recv_val = __shfl_xor_sync(mask, send_val, NUM_ENTRIES, SIZE);
+      (xor_bit_set ? CompileTimeArrayImpl<SIZE, idx>::val
+                   : CompileTimeArrayImpl<SIZE, idx + NUM_ENTRIES>::val) = recv_val;
+
+      constexpr int next_idx = idx + 1 + ((idx + 1) % NUM_ENTRIES == 0) * NUM_ENTRIES;
+      CompileTimeArrayImpl<SIZE, next_idx>::template Loop<NUM_ENTRIES>(xor_bit_set, mask);
+    }
+
+    // terminate recursion
+    template <int NUM_ENTRIES, std::enable_if_t<!NUM_ENTRIES, int> = 0>
```

Comment:
We tend to use `Int2Type` instead of `std::enable_if`. If you prefer `enable_if`, make sure not to use `_t` alias, since we have to support C++11. Also, prefer `::cuda::std`.

Reply:
In the [CHANGELOG.md](https://github.com/NVIDIA/cub/blob/main/CHANGELOG.md#deprecated-traits) it says that CUB nowadays targets C++14 and that the STL traits should be used? Do you want me to use the STL traits or `Int2Type`?

Reply:
We deprecated C++11, but the support is not removed yet. I'd prefer `Int2Type` here.